src/services/codefixes/convertToAsyncFunction.ts

```diff
@@ -0,0 +1,549 @@
+            else if (isPromiseReturningExpression(node, checker, "catch")) {
+                setOfExpressionsToReturn.set(getNodeId(node).toString(), true);
+                // if .catch() is the last call in the chain, move leftward in the chain until we hit something else that should be returned
+                forEachChild(node, visit);
+            }
+            else if (isPromiseReturningExpression(node, checker)) {
+                setOfExpressionsToReturn.set(getNodeId(node).toString(), true);
+                // don't recurse here, since we won't refactor any children or arguments of the expression
+            }
+            else {
+                forEachChild(node, visit);
+            }
+        });
+
+        return setOfExpressionsToReturn;
+    }
+
+    function isPromiseReturningExpression(node: Node, checker: TypeChecker, name?: string): boolean {
+        const isNodeExpression = name ? isCallExpression(node) : isExpression(node);
+        const isExpressionOfName = isNodeExpression && (!name || hasPropertyAccessExpressionWithName(node as CallExpression, name));
+        const nodeType = isExpressionOfName && checker.getTypeAtLocation(node);
+        return !!(nodeType && checker.getPromisedTypeOfPromise(nodeType));
+    }
+
+    function declaredInFile(symbol: Symbol, sourceFile: SourceFile): boolean {
+        return symbol.valueDeclaration && symbol.valueDeclaration.getSourceFile() === sourceFile;
+    }
+
+    /*
+        Renaming of identifiers may be neccesary as the refactor changes scopes -
+        This function collects all existing identifier names and names of identifiers that will be created in the refactor.
+        It then checks for any collisions and renames them through getSynthesizedDeepClone
+    */
+    function renameCollidingVarNames(nodeToRename: FunctionLikeDeclaration, checker: TypeChecker, synthNamesMap: Map<SynthIdentifier>, context: CodeFixContextBase, setOfAllExpressionsToReturn: Map<true>, originalType: Map<Type>, allVarNames: SymbolAndIdentifier[]): FunctionLikeDeclaration {
+
+        const identsToRenameMap: Map<Identifier> = createMap(); // key is the symbol id 
+        forEachChild(nodeToRename, function visit(node: Node) {
+            if (!isIdentifier(node)) {
+                forEachChild(node, visit);
+                return;
+            }
+
+            const symbol = checker.getSymbolAtLocation(node);
+            const isDefinedInFile = symbol && declaredInFile(symbol, context.sourceFile);
+
+            if (symbol && isDefinedInFile) {
+                const type = checker.getTypeAtLocation(node);
+                const callSignatures = type && type.getCallSignatures();
+                const symbolIdString = getSymbolId(symbol).toString();
+
+                // if the identifier refers to a function we want to add the new synthesized variable for the declaration (ex. blob in let blob = res(arg))
+                // Note - the choice of the first call signature is arbitrary

```

Comment:
Given a list of call signatures where you don't know which to pick, it's generally best in TypeScript to pick the last rather than the first.